src/jit/PerfDump.cpp

```diff
@@ -0,0 +1,161 @@
+    JIT_CODE_UNWINDING_INFO = 4 // record describing a function unwinding information
+};
+
+PerfDump& PerfDump::instance()
+{
+    static PerfDump instance;
+    return instance;
+}
+
+void PerfDump::openFile()
+{
+    if (this->m_file != nullptr) {
+        return;
+    }
+
+    this->m_pid = (uint32_t)getpid();
+    this->m_codeLoadIndex = 0;
+
+    std::string tmpName = "./jit-" + std::to_string(m_pid) + ".dump";
+    this->m_file = fopen(tmpName.c_str(), "w");
+    this->dumpFileHeader();
+
+    int fd = open(tmpName.c_str(), O_RDONLY | O_CLOEXEC, 0);
+    mmap(NULL, 1, PROT_READ | PROT_EXEC, MAP_SHARED, fd, 0);
+    close(fd);
+}
+
+void PerfDump::dumpFileHeader()
+{
+    struct {
+        const uint32_t magic;
+        const uint32_t version;
+        const uint32_t totalSize;
+        const uint32_t elfMach;
+        const uint32_t pad;
+        const uint32_t pid;
+        const uint64_t timestamp;
+        const uint64_t flags;
+    } fileHeader = {
+        .magic = DUMP_MAGIC,
+        .version = DUMP_VERSION,
+        .totalSize = sizeof(uint32_t) * 6 + sizeof(uint64_t) * 2,
+        .elfMach = ELFMACH,
+        .pad = 0,
+        .pid = m_pid,
+        .timestamp = (uint64_t)std::time(0),
+        .flags = 0
+    };
+
+    fwrite(&fileHeader, sizeof(fileHeader), 1, m_file);
+}
+
+void PerfDump::dumpRecordHeader(const uint32_t recordType, const uint32_t entrySize)
+{
+    struct {
+        const uint32_t recordType;
+        const uint32_t size;
+        const uint64_t timestamp;
+    } recordHeader = {
+        .recordType = recordType,
+        .size = entrySize + ((uint32_t)sizeof(uint32_t)) * 2 + (uint32_t)sizeof(uint64_t),
+        .timestamp = (uint64_t)std::time(0)
+    };
+
+    fwrite(&recordHeader, sizeof(recordHeader), 1, m_file);
+}
+
+void PerfDump::dumpCodeLoad(const uint64_t vma, const uint64_t codeAddr, const uint64_t codeSize, const std::string& functionName, const uint8_t* nativeCode)
+{
+    struct {
+        const uint32_t pid;
+        const uint32_t tid;
+        const uint64_t vma;
+        const uint64_t codeAddr;
+        const uint64_t codeSize;
+        const uint64_t codeLoadIndex;
+    } codeLoad1 = {
+        .pid = m_pid,
+        .tid = (uint32_t)std::hash<std::thread::id>{}(std::this_thread::get_id()),
+        .vma = vma,
+        .codeAddr = codeAddr,
+        .codeSize = codeSize,
+        .codeLoadIndex = m_codeLoadIndex++
+    };
+    const size_t nameSize = functionName.size() + 1;
+    dumpRecordHeader(JIT_CODE_LOAD, sizeof(codeLoad1) + nameSize + codeSize);
+    fwrite(&codeLoad1, sizeof(codeLoad1), 1, m_file);
+    fwrite(functionName.c_str(), nameSize, 1, m_file);
+    fwrite(nativeCode, codeSize, 1, m_file);
+}
+
+void PerfDump::dumpCodeClose()
+{
+    dumpRecordHeader(JIT_CODE_CLOSE, 0);
+}
+
+PerfDump::~PerfDump()
+{
+    if (this->m_file != nullptr) {
+        this->dumpCodeClose();
+        fclose(this->m_file);
```

Comment:
Currently, the dump file would be opened and maintained until the end of Walrus execution.
I'm wondering that this file should be opened during the execution of Walrus engine.
What about closing the file right after dump done?

Reply:
Compiling functions can be called any number of times, and we don't know when it is the last time when it is called. This feature is diagnostic only, which should never be enabled in production (even the default is disabled), so we hope having an extra opened file should not cause any issues.

We can also repopen the file every time before dump if that is better.

Reply:
Currently we use a single static instance for dumping, which is not thread friendly, but we can improve this with mutexes if it is needed later.

Shall we do any change with the file?

Reply:
I see, multi-thread support is not a concern now, so please ignore it.
one more thing about file open.
PerDump is globally instantiated once, so how about opening the dump file in the constructor of PerfDump similar to file close in the destructor?

Reply:
Good idea. We do that.